File: algorithms/2_medium/M/03853_merge_close_characters/stack.c
typedef ptrdiff_t isize;

char* mergeCharacters(const char* s, const int k_int) {
    if (s == NULL) { return NULL; }
    const isize len = (isize)strlen(s);
    const isize k = k_int;
    isize prev_idx[256];
    for (isize i = 0; i < 256; i += 1) { prev_idx[i] = -1; }
    char* stack = malloc((size_t)(len + 1) * sizeof(*stack));
    isize stack_len = 0;

    for (isize i = 0; i < len; i += 1) {
        const unsigned char ch = (unsigned char)s[i];
        if (prev_idx[ch] == -1 || stack_len - prev_idx[ch] > k) {
            prev_idx[ch] = stack_len;
            stack[stack_len] = ch;
            stack_len += 1;
        }
    }

    stack[stack_len] = '\0';
    return stack;
}

File: algorithms/2_medium/M/03853_merge_close_characters/stack.rs
impl Solution {
    pub fn merge_characters(s: String, k: i32) -> String {
        let k = k as isize;
        let mut prev_idx = [-1_isize; 256];
        let mut stack: Vec<u8> = Vec::with_capacity(s.len());

        for &b in s.as_bytes() {
            if prev_idx[b as usize] == -1 || (stack.len() as isize) - prev_idx[b as usize] > k {
                prev_idx[b as usize] = stack.len() as isize;
                stack.push(b);
            }
        }

        String::from_utf8(stack).unwrap()
    }
}
